File: python/xgrammar/xgrammar.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Classes handling the grammar guided generation of MLC LLM serving"""

from typing import List, Optional, Tuple, Union

import torch
from transformers import PreTrainedTokenizerBase

from . import xgrammar_bindings as _core


def _init_object_with_handle(type, handle):
    """Initialize an object with a handle."""
    obj = type.__new__(type)
    obj._handle = handle
    return obj


class BNFGrammar:
    """This class stores the abstract syntax tree (AST) of the Backus-Naur Form (BNF) grammar and
    provides utilities to parse and print the AST. User should provide a BNF/EBNF (Extended
    Backus-Naur Form) grammar, and use from_ebnf_string to parse and simplify the grammar into an
    AST of BNF grammar.
    """

    def __init__(self, ebnf_string: str, main_rule: str = "main") -> None:
        r"""Construct a BNF grammar with a EBNF-formatted string. The grammar will be normalized
        (simplified) by default.

        EBNF grammar: see https://www.w3.org/TR/xml/#sec-notation. Note:
        1. Use # as the comment mark
        2. Use C-style unicode escape sequence \u01AB, \U000001AB, \xAB
        3. A-B (match A and not match B) is not supported yet
        4. Lookahead assertion can be added at the end of a rule to speed up matching. E.g.
        ```
        main ::= "ab" a [a-z]
        a ::= "cd" (=[a-z])
        ```
        The assertion (=[a-z]) means a must be followed by [a-z].

        Parameters
        ----------
        ebnf_string : str
            The grammar string.

        main_rule : str
            The name of the main rule. Default: "main".

        Returns
        -------
        grammar : BNFGrammar
            The parsed BNF grammar.
        """
        self._handle = _core.BNFGrammar(ebnf_string, main_rule)

    def to_string(self) -> str:
        """Print the BNF grammar to a string, in standard BNF format.

        Returns
        -------
        grammar_string : str
            The BNF grammar string.
        """
        return self._handle.to_string()

    def __str__(self) -> str:
        return self.to_string()

    def serialize(self, prettify: bool = False) -> str:
        """Serialize the AST. Dump the raw representation of the AST to a JSON file.

        Parameters
        ----------
        prettify : bool
            Whether to format the JSON string. If False, all whitespaces will be removed.

        Returns
        -------
        json_string : str
            The JSON string.
        """
        return self._handle.serialize(prettify)

    @staticmethod
    def deserialize(json_string: str) -> "BNFGrammar":
        """Load a BNF grammar from the raw representation of the AST in JSON format.

        Parameters
        ----------
        json_string : str
            The JSON string.

        Returns
        -------
        grammar : BNFGrammar
            The loaded BNF grammar.
        """
        return _init_object_with_handle(BNFGrammar, _core.BNFGrammar.deserialize(json_string))

    @staticmethod
    def _init_no_normalization(
        ebnf_string: str,
        main_rule: str = "main",
    ) -> "BNFGrammar":
        r"""Construct a BNF grammar with a EBNF-formatted string, but not normalize it.
        For test purposes.

        Parameters
        ----------
        ebnf_string : str
            The grammar string.

        main_rule : str
            The name of the main rule. Default: "main".

        Returns
        -------
        grammar : BNFGrammar
            The parsed BNF grammar.
        """
        return _init_object_with_handle(
            BNFGrammar, _core.BNFGrammar._init_no_normalization(ebnf_string, main_rule)
        )


class BuiltinGrammar:
    @staticmethod
    def json() -> BNFGrammar:
        """Get the grammar of standard JSON.

        Returns
        -------
        grammar : BNFGrammar
            The JSON grammar.
        """
        return _init_object_with_handle(BNFGrammar, _core.BuiltinGrammar.json())

    @staticmethod
    def json_schema(
        schema: str,
        *,
        indent: Optional[int] = 2,
        separators: Optional[Tuple[str, str]] = None,
        strict_mode: bool = True
    ) -> BNFGrammar:
        """Construct a BNF grammar from the json schema string. The schema string should be in the
        format of the schema of a JSON file. We will parse the schema and generate a BNF grammar.

        Parameters
        ----------
        schema : str
            The schema string.

        indent : Optional[int]
            The number of spaces for indentation. If None, the output will be in one line.
            Default: None.

        separators : Optional[Tuple[str, str]]
            Two separators used in the schema: comma and colon. Examples: (",", ":"), (", ", ": ").
            If None, the default separators will be used: (",", ": ") when the indent is not None,
            and (", ", ": ") otherwise. This follows the convention in json.dumps(). Default: None.

        strict_mode : bool
            Whether to use strict mode. In strict mode, the generated grammar will not allow
            properties and items that is not specified in the schema. This is equivalent to
            setting unevaluatedProperties and unevaluatedItems to false.

            This helps LLM to generate accurate output in the grammar-guided generation with JSON
            schema. Default: True.

        Returns
        -------
        grammar : BNFGrammar
            The generated BNF grammar.
        """
        return _init_object_with_handle(
            BNFGrammar,
            _core.BuiltinGrammar.json_schema(schema, indent, separators, strict_mode),
        )

    @staticmethod
    def _json_schema_to_ebnf(
        schema: str,
        *,
        indent: Optional[int] = 2,
        separators: Optional[Tuple[str, str]] = None,
        strict_mode: bool = True
    ) -> str:
        """Convert JSON schema string to EBNF grammar string. For test purposes.

        Parameters
        ----------
        json_schema : str
            The JSON schema string.

        indent : Optional[int]
            The number of spaces for indentation. If None, the output will be in one line.
            Default: 2.

        separators : Optional[Tuple[str, str]]
            Two separators used in the schema: comma and colon. Examples: (",", ":"), (", ", ": ").
            If None, the default separators will be used: (",", ": ") when the indent is not None,
            and (", ", ": ") otherwise. This follows the convention in json.dumps(). Default: None.

        strict_mode : bool
            Whether to use strict mode. In strict mode, the generated grammar will not allow
            properties and items that is not specified in the schema. This is equivalent to
            setting unevaluatedProperties and unevaluatedItems to false.

            This helps LLM to generate accurate output in the grammar-guided generation with JSON
            schema. Default: True.

        Returns
        -------
        ebnf_string : str
            The EBNF grammar string.
        """
        return _core.BuiltinGrammar._json_schema_to_ebnf(schema, indent, separators, strict_mode)


class GrammarStateMatcher:
    """A stateful matcher to match tokens to the specified BNF grammar. This class is the core logic
    of the grammar-guided generation.

    This class implements the non-deterministic pushdown automaton (NPDA) matching algorithm to
    match characters to a BNF grammar. It keep track of the current state of the matching process by
    maintaining several stacks internally as possible paths in the NPDA. It also supports
    backtracking.

    It is particularly capable of finding the set of tokens that are acceptable for the next step
    and storing them in a bitmask. This aids in grammar-guided generation.

    Parameters
    ----------
    grammar : BNFGrammar
        The BNF grammar to match.

    tokenizer : Union[None, Tokenizer, List[str]]
        The tokenizer to use, or the list of tokens.

        (For debug purpose) If None, the matcher will use an empty token set, and can only accept
        and match characters. Default: None.

    max_rollback_steps : int
        The maximum number of steps to rollback when backtracking. Default: 0.
    """

    def __init__(
        self,
        grammar: BNFGrammar,
        tokenizer: Union[None, PreTrainedTokenizerBase, List[str]] = None,
        max_rollback_steps: int = 0,
    ):
        if isinstance(tokenizer, list):
            self.__init_handle_by_constructor__(
                _ffi_api.GrammarStateMatcherFromTokenTable,  # type: ignore  # pylint: disable=no-member
                grammar,
                tokenizer,
                max_rollback_steps,
            )
        else:
            self.__init_handle_by_constructor__(
                _ffi_api.GrammarStateMatcherFromTokenizer,  # type: ignore  # pylint: disable=no-member
                grammar,
                tokenizer,
                max_rollback_steps,
            )

    def accept_token(self, token_id: int, verbose: bool = False) -> bool:
        """Accept one token and update the state of the matcher.

        Parameters
        ----------
        token_id : int
            The id of the token to accept.

        Returns
        -------
        accepted : bool
            Whether the token is accepted.

        Note
        ----
        Termination state.

        When the end of the main rule is reached, the matcher can only accept the stop token.
        The matcher is terminated after accepting the stop token, i.e. no accept_token or
        find_next_rejected_tokens operations can be performed. The termination state can be canceled
        using Rollback().
        """
        return _ffi_api.GrammarStateMatcherAcceptToken(self, token_id)  # type: ignore  # pylint: disable=no-member

    def _accept_string(self, input_str: str, verbose: bool = False) -> bool:
        """Accept one unicode codepoint to the current state. For test purposes.

        Parameters
        ----------
        codepoint : int
            The unicode codepoint of the character to be accepted.
        """
        return _ffi_api.GrammarStateMatcherDebugAcceptChar(  # type: ignore  # pylint: disable=no-member
            self, codepoint, verbose
        )

    def find_next_token_bitmask(self, verbose: bool = False) -> torch.Tensor:
        """Find the ids of the rejected tokens for the next step.

        Parameters
        ----------
        verbose : bool
            Whether to print information about timing and result counts to stderr.
            For debug purposes. Default: False.

        Returns
        -------
        rejected_token_bitmask : torch.Tensor
            A tensor of rejected token ids.
        """
        return _ffi_api.GrammarStateMatcherFindNextRejectedTokenBitmask(self, verbose)

    @staticmethod
    def get_rejected_tokens_from_bitmask(bitmask: torch.Tensor) -> List[int]:
        """Get the ids of the rejected tokens from the bitmask.

        Parameters
        ----------
        bitmask : torch.Tensor
            The rejected token bitmask.

        Returns
        -------
        rejected_token_ids : List[int]
            A list of rejected token ids.
        """
        return None

    @staticmethod
    def apply_token_bitmask(tensor: torch.Tensor, bitmask: torch.Tensor) -> torch.Tensor:
        """Apply the bitmask to the tensor.

        Parameters
        ----------
        tensor : torch.Tensor
            The tensor to apply the bitmask to.

        bitmask : torch.Tensor
            The bitmask to apply.

        Returns
        -------
        masked_tensor : torch.Tensor
            The masked tensor.
        """
        return None

    def find_jump_forward_string(self) -> str:
        """Find the jump-forward string for jump-forward decoding. This is the longest string that
        will be valid according to the current syntax.

        Notes
        -----
        This method does not change the grammar state.

        Returns
        -------
        jump_forward_string : str
            The jump-forward string.
        """
        return _ffi_api.GrammarStateMatcherFindJumpForwardString(self)  # type: ignore  # pylint: disable=no-member

    def rollback(self, num_tokens: int) -> None:
        """Rollback the matcher to a previous state.

        Parameters
        ----------
        num_tokens : int
            The number of tokens to rollback. It cannot exceed the current number of steps, nor can
            it exceed the specified maximum number of rollback steps.
        """
        _ffi_api.GrammarStateMatcherRollback(self, num_tokens)  # type: ignore  # pylint: disable=no-member

    def max_rollback_steps(self) -> int:
        """Get the maximum number of rollback steps allowed.

        Returns
        -------
        max_rollback_steps : int
            The maximum number of rollback steps.
        """
        return _ffi_api.GrammarStateMatcherMaxRollbackSteps(self)  # type: ignore  # pylint: disable=no-member

    def is_terminated(self) -> bool:
        """Check if the matcher has accepted the stop token and terminated. See also
        GrammarStateMatcher.accept_token.

        Returns
        -------
        terminated : bool
            Whether the matcher has terminated.
        """
        return _ffi_api.GrammarStateMatcherIsTerminated(self)  # type: ignore  # pylint: disable=no-member

    def reset_state(self) -> None:
        """Reset the matcher to the initial state."""
        _ffi_api.GrammarStateMatcherResetState(self)  # type: ignore  # pylint: disable=no-member

File: web/src/xgrammar.ts
import Module from "./xgrammar_binding";

let binding: any = null;

async function asyncInitBinding() {
  if (binding == null) {
    binding = await Module();
  }
}

/**
 * This class stores the abstract syntax tree (AST) of the Backus-Naur Form (BNF) grammar and
 * provides utilities to parse and print the AST. User should provide a BNF/EBNF (Extended
 * Backus-Naur Form) grammar, and use from_ebnf_string to parse and simplify the grammar into an
 * AST of BNF grammar.
 */
export class BNFGrammar {
  handle: any;

  /**
   * @internal
   * Private constructor. Factory methods are used since binding initialization is asynchronous.
   * @param {any} handle handle of BNFGrammar created by binding.
   */
  constructor(handle: any) {
    this.handle = handle;
  }

  /**
   * Dispose this BNFGrammar.
   */
  dispose() {
    this.handle.delete();
  }

  /**
   * Construct a BNF grammar with a EBNF-formatted string. The grammar will be normalized
   * (simplified) by default.
   * EBNF grammar: see https://www.w3.org/TR/xml/#sec-notation. Note:
   * 1. Use # as the comment mark
   * 2. Use C-style unicode escape sequence \u01AB, \U000001AB, \xAB
   * 3. A-B (match A and not match B) is not supported yet
   * 4. Lookahead assertion can be added at the end of a rule to speed up matching. E.g.
   * ```
   * main ::= "ab" a [a-z]
   * a ::= "cd" (=[a-z])
   * ```
   * The assertion (=[a-z]) means a must be followed by [a-z].
   * @param {string} ebnfString The grammar string
   * @param {string} [mainRule="main"] The name of the main rule. Default: "main".
   * @returns {BNFGrammar} The parsed BNF grammar.
   */
  static async createBNFGrammar(ebnfString: string, mainRule = "main"): Promise<BNFGrammar> {
    await asyncInitBinding();
    return new BNFGrammar(new binding.BNFGrammar(ebnfString, mainRule));
  }

  /**
   * Load a BNF grammar from the raw representation of the AST in JSON format.
   * @param {string} json_string The JSON string.
   * @returns {BNFGrammar} The loaded BNF grammar.
   */
  static async deserialize(json_string: string): Promise<BNFGrammar> {
    await asyncInitBinding();
    return new BNFGrammar(binding.BNFGrammar.Deserialize(json_string));
  }

  /**
   * Print the BNF grammar to a string, in standard BNF format.
   * @returns The BNF grammar string.
   */
  toString(): string {
    return this.handle.ToString();
  }

  /**
   * Serialize the AST. Dump the raw representation of the AST to a JSON file.
   * @param {boolean} [prettify=false] Whether to format the JSON string. If False, all whitespaces
   * will be removed.
   * @returns {string} The JSON string
   */
  serialize(prettify = false): string {
    return this.handle.Serialize(prettify);
  }
}

export class BuiltinGrammar {
  /**
   * Get the grammar of standard JSON.
   * @returns {BNFGrammar} The JSON grammar.
   */
  static async json(): Promise<BNFGrammar> {
    await asyncInitBinding();
    return new BNFGrammar(new binding.BuiltinGrammar.JSON());
  }

  /**
   * Construct a BNF grammar from the json schema string. The schema string should be in the 
   * format of the schema of a JSON file. We will parse the schema and generate a BNF grammar.
   *
   * @param {string} schema The schema string.
   * @param {number} [indent=2] The number of spaces for indentation. If -1, the grammar will
   * enforce the output to be in one line.
   * @param {[string, string]} [separators] Two separators that will be enforced by the grammar:
   * comma and colon. Examples: (",", ":"), (", ", ": "). If undefined, the default separators will
   * be used: (",", ": ") when the indent is not undefined, and (", ", ": ") otherwise. This follows
   * the convention in Python's json.dumps(). Currently unsupported and will use the default value.
   * @param {boolean} [strictMode=true] Whether to use strict mode. In strict mode, the generated
   * grammar will not allow properties and items that is not specified in the schema. This is
   * equivalent to setting unevaluatedProperties and unevaluatedItems to false.
   * @returns {BNFGrammar} The generated BNF grammar.
   */
  static async jsonSchema(
    schema: string,
    indent = 2,
    separators?: [string, string],
    strictMode = true
  ): Promise<BNFGrammar> {
    // TODO(Charlie): Add support for separators, which requires binding std::pair
    // in emscripten
    if (separators !== undefined) {
      throw new Error(
        `Argument separators is not supported yet, please leave it as undefined, and the ` +
        `default value (",", ": ") will be used.`
      );
    }
    await asyncInitBinding();
    // indent being -1 is equivalent to not having a value for the std::optional arg in C++.
    // This is a workaround to Typescript not being able to express Optional value like Python; if
    // user specifies indent to be undefined, it still becomes 2.
    let optionalIndent: number | undefined = indent == -1 ? undefined : indent;
    return new BNFGrammar(
      new binding.BuiltinGrammar.JSONSchema(schema, optionalIndent, separators, strictMode));
  }

  /**
   * Convert JSON schema string to EBNF grammar string. For test purposes.
   *
   * @param {string} schema The schema string.
   * @param {number} [indent=2] The number of spaces for indentation. If -1, the grammar will
   * enforce the output to be in one line.
   * @param {[string, string]} [separators] Two separators that will be enforced by the grammar:
   * comma and colon. Examples: (",", ":"), (", ", ": "). If undefined, the default separators will
   * be used: (",", ": ") when the indent is not undefined, and (", ", ": ") otherwise. This follows
   * the convention in Python's json.dumps(). Currently unsupported and will use the default value.
   * @param {boolean} [strictMode=true] Whether to use strict mode. In strict mode, the generated
   * grammar will not allow properties and items that is not specified in the schema. This is
   * equivalent to setting unevaluatedProperties and unevaluatedItems to false.
   * @returns {string} The EBNF grammar string.
   */
  static async _jsonSchemaToEBNF(
    schema: string,
    indent = 2,
    separators?: [string, string],
    strictMode = true
  ): Promise<string> {
    // TODO(Charlie): Add support for separators, which requires binding std::pair
    // in emscripten
    if (separators !== undefined) {
      throw new Error(
        `Argument separators is not supported yet, please leave it as undefined, and the ` +
        `default value (",", ": ") will be used.`
      );
    }
    await asyncInitBinding();
    // indent being -1 is equivalent to not having a value for the std::optional arg in C++.
    // This is a workaround to Typescript not being able to express Optional value like Python; if
    // user specifies indent to be undefined, it still becomes 2.
    let optionalIndent: number | undefined = indent == -1 ? undefined : indent;
    return binding.BuiltinGrammar._JSONSchemaToEBNF(schema, optionalIndent, separators, strictMode);
  }
}

/**
 * A class that wraps a decoded token table, needed to instantiate GrammarStateMatcher.
 */
export class XGTokenTable {
  /** A handle to the decoded token table of type binding.VectorString. */
  decodedTokenTable: any;

  /**
   * @internal
   * Private constructor. Factory methods are used since binding initialization is asynchronous.
   * @param {any} decodedTokenTable Post-processed token table.
   */
  private constructor(decodedTokenTable: any) {
    this.decodedTokenTable = decodedTokenTable;
  };

  /**
   * Dispose this token table.
   */
  dispose() {
    this.decodedTokenTable.delete();
  }

  /**
   * Instantiate with raw token table and the decoder type by internally post-processing
   * the raw token table by decoding each token with the provided decoder type.
   * @param {string[]} rawTokenTable: the token table in the form of a string list of tokens,
   * ordered by their token id. It should include all the special tokens.
   * @param {string} decoderType: either "byte_fallback", or "byte_level". See `tokenizer.cc` for
   * its semantic.
   */
  static async createXGTokenTable(
    rawTokenTable: string[],
    decoderType: string,
  ): Promise<XGTokenTable> {
    await asyncInitBinding();
    // Convert string[] to std::vector<std::string>
    const rawTokenTableVec = binding.vecStringFromJSArray(rawTokenTable);
    // Returns of type binding.VectorString
    const decodedTokenTable = binding.DecodeTokenTable(rawTokenTableVec, decoderType);
    rawTokenTableVec.delete();
    // Instantiate XGTokenTable
    return new XGTokenTable(decodedTokenTable);
  }
}

/**
 * A stateful matcher to match tokens to the specified BNF grammar. This class is the core logic
 * of the grammar-guided generation.
 *
 * This class implements the non-deterministic pushdown automaton (NPDA) matching algorithm to
 * match characters to a BNF grammar. It keep track of the current state of the matching process by
 * maintaining several stacks internally as possible paths in the NPDA. It also supports
 * backtracking.
 *
 * It is particularly capable of finding the set of tokens that are acceptable for the next step
 * and storing them in a bitmask. This aids in grammar-guided generation.
 */
export class GrammarStateMatcher {
  private handle: any;

  /**
   * @internal
   * Private constructor. Factory methods are used since binding initialization is asynchronous.
   * @param {any} handle handle of GrammarStateMatcher created by binding.
   */
  private constructor(handle: any) {
    this.handle = handle;
  }

  /**
   * Dispose this grammar state matcher.
   */
  dispose() {
    this.handle.delete();
  }

  /**
   * Construct a GrammarStateMatcher.
   * @param {BNFGrammar} bnfGrammar The BNF grammar to match.
   * @param {XGTokenTable} tokenTable The decoded token table.
   * @param {number[] | number} [stopTokenIds=undefined] Stop tokens to override the default ones.
   * @param {boolean} [terminateWithoutStopToken=false] Whether to terminate without stop token.
   * @param {number} [maxRollbackSteps=0] Max rollback steps.
   * @returns {GrammarStateMatcher} The constructed GrammarStateMatcher.
   */
  static async createGrammarStateMatcher(
    bnfGrammar: BNFGrammar,
    tokenTable: XGTokenTable,
    stopTokenIds?: number[] | number,
    terminateWithoutStopToken: boolean = false,
    maxRollbackSteps: number = 0,
  ): Promise<GrammarStateMatcher> {
    await asyncInitBinding();
    // Convert stopTokenIds to std::vector<int> if not undefined
    if (stopTokenIds !== undefined) {
      if (!Array.isArray(stopTokenIds)) {
        stopTokenIds = [stopTokenIds];
      }
      stopTokenIds = binding.vecIntFromJSArray(stopTokenIds);
    }
    return new GrammarStateMatcher(new binding.GrammarStateMatcher(
      bnfGrammar.handle,
      tokenTable.decodedTokenTable,
      stopTokenIds,
      terminateWithoutStopToken,
      maxRollbackSteps,
    ));
  }

  /**
   * Get the vocab size.
   */
  getVocabSize(): number {
    return this.handle.GetVocabSize();
  }

  /**
   * Get the maximum number of rollback steps allowed.
   */
  getMaxRollbackSteps(): number {
    return this.handle.GetMaxRollbackSteps();
  }

  /**
   * Accept one token and update the state of the matcher.
   * @param {number} tokenID The id of the token to accept.
   * @param {boolean} [verbose=false] To print debugging info
   * @returns {boolean} Whether the token is accepted.
   */
  acceptToken(tokenID: number, verbose: boolean = false): boolean {
    return this.handle.AcceptToken(tokenID, verbose);
  }

  /**
   * Accept one unicode codepoint to the current state. For test purposes.
   * @param {string} inputStr The unicode codepoint of the character to be accepted.
   * @param {boolean} [verbose=false] To print debugging info
   * @returns {boolean} Whether the input string is accepted.
   */
  _acceptString(inputStr: string, verbose: boolean = false): boolean {
    return this.handle._AcceptString(inputStr, verbose);
  }

  /**
   * Returns a bitmask in the form of an Int32Array of length ceildiv(vocab_size, 32)
   * based on what tokens can/cannot be accepted by the current state of the grammar state matcher.
   *
   * @returns {Int32Array} An array representing the bitmask that masks the rejected token IDs
   */
  async findNextTokenBitmask(): Promise<Int32Array> {
    await asyncInitBinding();
    const maskIntVector = this.handle.FindNextTokenBitmask()  // a handle of std::vector<int32_t>
    const maskInt32Array = binding.vecIntToView(maskIntVector).slice();
    maskIntVector.delete();
    return maskInt32Array;
  }

  /**
   * 
   * @param {Int32Array} bitmask Bitmask returned by findNextTokenBitmask().
   * @param {number} vocabSize Vocab size returned by getVocabSize().
   * @returns An array of vocab ID that will be rejected as a result of the bitmask.
   */
  static async getRejectedTokensFromBitmask(
    bitmask: Int32Array,
    vocabSize: number
  ): Promise<Int32Array> {
    await asyncInitBinding();
    const bitmaskIntVector = binding.vecIntFromJSArray(bitmask);
    const rejectedIDsIntVector = binding.GrammarStateMatcher.GetRejectedTokensFromBitMask(
      bitmaskIntVector,
      vocabSize
    );
    bitmaskIntVector.delete();
    const rejectedIDsInt32Array = binding.vecIntToView(rejectedIDsIntVector).slice();
    rejectedIDsIntVector.delete();
    return rejectedIDsInt32Array;
  }

  /**
   * Check if the matcher has accepted the stop token and terminated. See also
   * GrammarStateMatcher.acceptToken.
   */
  isTerminated(): boolean {
    return this.handle.IsTerminated();
  }

  /**
   * Reset the matcher to the initial state.
   */
  reset(): void {
    this.handle.Reset();
  }

  /**
   * Find the jump-forward string for jump-forward decoding. This is the longest string that
   * will be valid according to the current syntax.
   * @returns {string} The jump-forward string.
   */
  findJumpForwardString(): string {
    return this.handle.FindJumpForwardString();
  }

  /**
   * Rollback the matcher to a previous state.
   * @param {number} numTokens The number of tokens to rollback. It cannot exceed the current
   * number of steps, nor can it exceed the specified maximum number of rollback steps.
   */
  rollBack(numTokens: number): void {
    this.handle.Rollback(numTokens);
  }
}
